edge/pkg/devicetwin/dtmanager/dmiserver.go

```diff
@@ -0,0 +1,361 @@
+	pb "github.com/kubeedge/kubeedge/edge/pkg/apis/dmi/v1"
+	"github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dmiclient"
+	"github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dmiserver"
+	"github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dtcommon"
+	"github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dtcontext"
+	"github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dttype"
+	"github.com/kubeedge/kubeedge/edge/pkg/metamanager/dao"
+	"github.com/kubeedge/kubeedge/pkg/apis/devices/v1alpha2"
+)
+
+// ResourceTypeDeviceModel is type for device model distribute to node
+const (
+	ResourceTypeDeviceModel  = "devicemodel"
+	ResourceTypeDevice       = "device"
+	ResourceTypeDeviceMapper = "devicemapper"
+)
+
+//TwinWorker deal twin event
+type DMIWorker struct {
+	Worker
+	Group string
+}
+
+var (
+	//dmiActionCallBack map for action to callback
+	dmiActionCallBack map[string]CallBack
+
+	MapperList        map[string]*pb.MapperInfo
+	DeviceModelList   map[string]*v1alpha2.DeviceModel
+	DeviceList        map[string]*v1alpha2.Device
+	PendingDeviceList map[string]*v1alpha2.Device
+)
+
+//Start worker
+func (dw DMIWorker) Start() {
+	klog.Infoln("dmi worker start")
+	initDMIActionCallBack()
+	initDeviceModelInfoFromDB()
+	initDeviceInfoFromDB()
+	initDeviceMapperInfoFromDB()
+	initPendingDeviceList()
+
+	go dmiserver.StartDMIServer(MapperList, DeviceList, DeviceModelList)
+	go dealPendingDeviceList()
+
+	for {
+		select {
+		case msg, ok := <-dw.ReceiverChan:
+			if !ok {
+				return
+			}
+
+			if dtMsg, isDTMessage := msg.(*dttype.DTMessage); isDTMessage {
+				if fn, exist := dmiActionCallBack[dtMsg.Action]; exist {
+					err := fn(dw.DTContexts, dtMsg.Identity, dtMsg.Msg)
+					if err != nil {
+						klog.Errorf("DMIModule deal %s event failed: %v", dtMsg.Action, err)
+					}
+				} else {
+					klog.Errorf("DMIModule deal %s event failed, not found callback", dtMsg.Action)
+				}
+			}
+
+		case v, ok := <-dw.HeartBeatChan:
+			if !ok {
+				return
+			}
+			if err := dw.DTContexts.HeartBeat(dw.Group, v); err != nil {
+				return
+			}
+		}
+	}
+}
+
+func initDMIActionCallBack() {
+	dmiActionCallBack = make(map[string]CallBack)
+	dmiActionCallBack[dtcommon.MetaDeviceOperation] = dealMetaDeviceOperation
+}
+
+func dealMetaDeviceOperation(context *dtcontext.DTContext, resource string, msg interface{}) error {
+	message, ok := msg.(*model.Message)
+	if !ok {
+		return errors.New("msg not Message type")
+	}
+	resources := strings.Split(message.Router.Resource, "/")
+	if len(resources) != 3 {
+		return fmt.Errorf("wrong resources %s", message.Router.Resource)
+	}
+	var device v1alpha2.Device
+	var dm v1alpha2.DeviceModel
+	switch resources[1] {
+	case "device":
+		switch message.GetOperation() {
+		case "insert":
+			err := json.Unmarshal(message.Content.([]byte), &device)
+			if err != nil {
+				return fmt.Errorf("invalid message content with err: %+v", err)
+			}
+
+			dm, ok := DeviceModelList[device.Spec.DeviceModelRef.Name]
```

Comment:
does `DeviceModelList` use in other package? It is best not to capitalize the first letter

Reply:
ok. these 4 list are all adjusted.